internal/controller/nicconfigurationtemplate_controller_test.go

```diff
@@ -0,0 +1,386 @@
+
+		device1 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device1", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device1)).To(Succeed())
+		device1.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:3b:00.0"}},
+			SerialNumber: "serialNumber1",
+		}
+		Expect(k8sClient.Status().Update(ctx, device1)).To(Succeed())
+
+		device2 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device2", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device2)).To(Succeed())
+		device2.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:d8:00.0"}},
+			SerialNumber: "serialNumber2",
+		}
+		Expect(k8sClient.Status().Update(ctx, device2)).To(Succeed())
+
+		// device3 doesn't match node selector
+		device3 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device3", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device3)).To(Succeed())
+		device3.Status = v1alpha1.NicDeviceStatus{
+			Node:         invalidNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:3b:00.0"}},
+			SerialNumber: "serialNumber1",
+		}
+		Expect(k8sClient.Status().Update(ctx, device3)).To(Succeed())
+
+		// device4 doesn't match nic type selector
+		device4 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device4", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device4)).To(Succeed())
+		device4.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX4",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:d8:00.0"}},
+			SerialNumber: "serialNumber1",
+		}
+		Expect(k8sClient.Status().Update(ctx, device4)).To(Succeed())
+
+		// device5 doesn't match PCI addresses selector
+		device5 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device5", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device5)).To(Succeed())
+		device5.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:81:00.0"}},
+			SerialNumber: "serialNumber1",
+		}
+		Expect(k8sClient.Status().Update(ctx, device5)).To(Succeed())
+
+		// device6 doesn't match Serial numbers selector
+		device6 := &v1alpha1.NicDevice{ObjectMeta: metav1.ObjectMeta{Name: "device6", Namespace: namespaceName}}
+		Expect(k8sClient.Create(ctx, device6)).To(Succeed())
+		device6.Status = v1alpha1.NicDeviceStatus{
+			Node:         validNode.Name,
+			Type:         "ConnectX6",
+			Ports:        []v1alpha1.NicDevicePortSpec{{PCI: "0000:3b:00.0"}},
+			SerialNumber: "serialNumber3",
+		}
+		Expect(k8sClient.Status().Update(ctx, device6)).To(Succeed())
+
+		Expect(k8sClient).To(Not(BeNil()))
```

Comment:
nit: it looks like it is a bit too late to check this here :) You already used the client multiple times

Reply:
You're right :) Dropped this check, it's not needed actually